File: experiment-5/runs/language=typescript_model=claude-opus-4-8_tooling=none/rep3/tests/queries.test.ts
/**
 * ============================================================================
 * Context: BDD tests — Query Engine
 * ----------------------------------------------------------------------------
 * Feature : The five capability areas from the spec — match search, team
 *           statistics, head-to-head, player search, standings and aggregate
 *           analysis. Each scenario is Given/When/Then.
 * ============================================================================
 */

import { describe, it, expect, beforeAll } from "vitest";
import { loadDataset, type Dataset } from "../src/dataLoader.js";
import {
  aggregateStats,
  biggestWins,
  competitionsForTeam,
  headToHead,
  rankTeams,
  searchMatches,
  searchPlayers,
  standings,
  teamStats,
} from "../src/queries.js";

let data: Dataset;
beforeAll(() => {
  data = loadDataset();
});

describe("Feature: Match Queries", () => {
  it("Scenario: find matches between two teams", () => {
    // Given the match data is loaded
    // When I search for matches between Flamengo and Fluminense
    const result = searchMatches(data.matches, { team: "Flamengo", opponent: "Fluminense" });
    // Then I receive a non-empty list
    expect(result.length).toBeGreaterThan(0);
    // And every match involves exactly those two teams
    for (const m of result) {
      const teams = [m.homeTeam, m.awayTeam].map((t) => t.toLowerCase());
      const hasFla = teams.some((t) => t.includes("flamengo"));
      const hasFlu = teams.some((t) => t.includes("fluminense"));
      expect(hasFla && hasFlu).toBe(true);
      // And each match has date and scores
      expect(m.competition).toBeTruthy();
    }
  });

  it("Scenario: results are sorted by date descending", () => {
    const result = searchMatches(data.matches, { team: "Palmeiras", competition: "Série A" });
    const dated = result.filter((m) => m.date).map((m) => m.date as string);
    const sorted = [...dated].sort().reverse();
    expect(dated).toEqual(sorted);
  });

  it("Scenario: filter matches by competition and season", () => {
    const result = searchMatches(data.matches, { competition: "Libertadores", season: 2014 });
    expect(result.length).toBeGreaterThan(0);
    for (const m of result) {
      expect(m.competition).toBe("Copa Libertadores");
      expect(m.season).toBe(2014);
    }
  });

  it("Scenario: restrict a team to away fixtures only", () => {
    const result = searchMatches(data.matches, {
      team: "Santos",
      awayOnly: true,
      competition: "Série A",
    });
    expect(result.length).toBeGreaterThan(0);
    for (const m of result) {
      expect(m.awayTeam.toLowerCase()).toContain("santos");
    }
  });
});

describe("Feature: Team Queries", () => {
  it("Scenario: get team statistics for a season", () => {
    // When I request statistics for Palmeiras in a season
    const record = teamStats(data.matches, { team: "Palmeiras", season: 2018 });
    // Then I receive wins, losses, draws and goals
    expect(record.matches).toBeGreaterThan(0);
    expect(record.wins + record.draws + record.losses).toBe(record.matches);
    expect(record.points).toBe(record.wins * 3 + record.draws);
    expect(record.goalsFor).toBeGreaterThanOrEqual(0);
  });

  it("Scenario: home record differs from overall record", () => {
    const home = teamStats(data.matches, { team: "Corinthians", venue: "home", competition: "Série A" });
    const all = teamStats(data.matches, { team: "Corinthians", competition: "Série A" });
    expect(home.matches).toBeLessThanOrEqual(all.matches);
    expect(home.matches).toBeGreaterThan(0);
  });

  it("Scenario: head-to-head totals are internally consistent", () => {
    const h2h = headToHead(data.matches, "Palmeiras", "Santos");
    expect(h2h.matches.length).toBeGreaterThan(0);
    const decided = h2h.teamAWins + h2h.teamBWins + h2h.draws;
    // decided counts only matches with a known score; never exceeds total meetings
    expect(decided).toBeLessThanOrEqual(h2h.matches.length);
  });

  it("Scenario: list competitions a team has played in", () => {
    const comps = competitionsForTeam(data.matches, "Flamengo");
    expect(comps.length).toBeGreaterThan(1);
  });
});

describe("Feature: Player Queries", () => {
  it("Scenario: search a player by name", () => {
    // When I look up Gabriel Barbosa / Neymar
    const neymar = searchPlayers(data.players, { name: "Neymar" });
    expect(neymar.length).toBeGreaterThan(0);
    expect(neymar[0].name.toLowerCase()).toContain("neymar");
  });

  it("Scenario: find Brazilian players sorted by rating", () => {
    const brazilians = searchPlayers(data.players, { nationality: "Brazil", limit: 10 });
    expect(brazilians.length).toBe(10);
    // Then results are sorted by overall descending
    for (let i = 1; i < brazilians.length; i++) {
      expect(brazilians[i - 1].overall ?? 0).toBeGreaterThanOrEqual(brazilians[i].overall ?? 0);
    }
    for (const p of brazilians) expect(p.nationality).toBe("Brazil");
  });

  it("Scenario: filter players by club and minimum rating", () => {
    const result = searchPlayers(data.players, { club: "Flamengo", minOverall: 70 });
    for (const p of result) {
      expect(p.club.toLowerCase()).toContain("flamengo");
      expect(p.overall ?? 0).toBeGreaterThanOrEqual(70);
    }
  });
});

describe("Feature: Competition Queries", () => {
  it("Scenario: calculate standings for a Brasileirão season", () => {
    // The 2019 Brasileirão was a 20-team, 38-round league.
    const table = standings(data.matches, "Série A", 2019);
    expect(table.length).toBeGreaterThanOrEqual(18);
    // Position 1 has the most (or tied-most) points
    expect(table[0].position).toBe(1);
    for (let i = 1; i < table.length; i++) {
      expect(table[i - 1].points).toBeGreaterThanOrEqual(table[i].points);
    }
  });

  it("Scenario: a known champion tops the table", () => {
    // Flamengo won the 2019 Brasileirão. Verify they finish first.
    const table = standings(data.matches, "Série A", 2019);
    expect(table[0].team.toLowerCase()).toContain("flamengo");
  });
});

describe("Feature: Statistical Analysis", () => {
  it("Scenario: average goals per match is plausible", () => {
    const stats = aggregateStats(searchMatches(data.matches, { competition: "Série A" }));
    expect(stats.goalsPerMatch).toBeGreaterThan(1.5);
    expect(stats.goalsPerMatch).toBeLessThan(4);
    // win-rate components sum to ~1
    const sum = stats.homeWinRate + stats.awayWinRate + stats.drawRate;
    expect(sum).toBeGreaterThan(0.99);
    expect(sum).toBeLessThan(1.01);
  });

  it("Scenario: biggest wins are sorted by margin", () => {
    const wins = biggestWins(searchMatches(data.matches, { competition: "Série A" }), 5);
    expect(wins.length).toBe(5);
    const margins = wins.map((m) => Math.abs((m.homeGoals as number) - (m.awayGoals as number)));
    for (let i = 1; i < margins.length; i++) {
      expect(margins[i - 1]).toBeGreaterThanOrEqual(margins[i]);
    }
    expect(margins[0]).toBeGreaterThanOrEqual(4);
  });

  it("Scenario: rank teams by wins in a season", () => {
    const ranking = rankTeams(data.matches, {
      competition: "Série A",
      season: 2019,
      metric: "wins",
      limit: 5,
    });
    expect(ranking.length).toBe(5);
    for (let i = 1; i < ranking.length; i++) {
      expect(ranking[i - 1].value).toBeGreaterThanOrEqual(ranking[i].value);
    }
  });
});

File: experiments/adrianco/experiment-7/brazil/runs/language=clojure_model=claude-opus-4-8-fast/rep3/test/brazilian_soccer/queries_test.clj
;; =============================================================================
;; brazilian-soccer.queries-test
;; -----------------------------------------------------------------------------
;; BDD (Gherkin-style Given/When/Then) coverage of the five capability groups
;; required by TASK.md: match, team, player, competition and statistical queries.
;; =============================================================================
(ns brazilian-soccer.queries-test
  (:require [clojure.test :refer [deftest testing is]]
            [brazilian-soccer.queries :as q]
            [brazilian-soccer.test-helper :as h]))

;; ---- 1. Match queries -----------------------------------------------------

(deftest find-matches-between-two-teams
  (testing "Given match data, When searching Flamengo vs Fluminense, Then we get dated, scored matches"
    (let [ms (q/find-matches (h/graph) {:team "Flamengo" :opponent "Fluminense"})]
      (is (seq ms))
      (is (every? :date ms))
      (is (every? #(or (= "Flamengo" (:home %)) (= "Flamengo" (:away %))) ms))
      (is (every? #(or (= "Fluminense" (:home %)) (= "Fluminense" (:away %))) ms)))))

(deftest find-matches-by-season-and-team
  (testing "Given a season filter, When searching Palmeiras in 2019, Then only 2019 matches return"
    (let [ms (q/find-matches (h/graph) {:team "Palmeiras" :season 2019})]
      (is (seq ms))
      (is (every? #(= 2019 (:season %)) ms)))))

(deftest find-matches-by-competition
  (testing "Given a competition filter, When searching Libertadores, Then all results are that competition"
    (let [ms (q/find-matches (h/graph) {:competition "Libertadores" :limit 100})]
      (is (seq ms))
      (is (every? #(= "Copa Libertadores" (:competition %)) ms)))))

(deftest find-matches-home-filter
  (testing "Given a home-only filter, When searching Corinthians at home, Then it is always the home side"
    (let [ms (q/find-matches (h/graph) {:team "Corinthians" :home true :season 2022})]
      (is (seq ms))
      (is (every? #(= "Corinthians" (:home %)) ms)))))

;; ---- 2. Team queries ------------------------------------------------------

(deftest team-stats-test
  (testing "Given match data, When computing Palmeiras 2019 stats, Then W/D/L and goals add up"
    (let [s (q/team-stats (h/graph) {:team "Palmeiras" :season 2019})]
      (is (some? s))
      (is (pos? (:matches s)))
      (is (= (:matches s) (+ (:wins s) (:draws s) (:losses s))))
      (is (>= (:goals-for s) 0)))))

(deftest head-to-head-test
  (testing "Given two rivals, When computing head-to-head, Then totals are internally consistent"
    (let [h (q/head-to-head (h/graph) "Flamengo" "Fluminense")]
      (is (pos? (:total h)))
      (is (= (:total h) (count (:matches h))))
      (is (= (:total h) (+ (:a-wins h) (:b-wins h) (:draws h)))))))

;; ---- 3. Player queries ----------------------------------------------------

(deftest brazilian-players-test
  (testing "Given player data, When filtering by nationality Brazil, Then only Brazilians return, rating-sorted"
    (let [ps (q/find-players (h/graph) {:nationality "Brazil" :limit 20})]
      (is (seq ps))
      (is (every? #(= "Brazil" (:nationality %)) ps))
      (is (apply >= (map :overall ps)) "sorted by overall descending"))))

(deftest player-by-name-test
  (testing "Given a name search, When looking up a known player, Then a match returns"
    (let [ps (q/find-players (h/graph) {:name "Neymar"})]
      (is (seq ps))
      (is (some #(re-find #"(?i)neymar" (:name %)) ps)))))

(deftest players-by-position-test
  (testing "Given a position filter, When searching GK, Then every result is a GK"
    (let [ps (q/find-players (h/graph) {:position "GK" :nationality "Brazil" :limit 10})]
      (is (seq ps))
      (is (every? #(= "GK" (:position %)) ps)))))

;; ---- 4. Competition queries ----------------------------------------------

(deftest standings-test
  (testing "Given a league season, When computing the 2019 Brasileirão table, Then Flamengo top a sane 20-team table"
    (let [rows (q/standings (h/graph) "Brasileirão Série A" 2019)]
      (is (seq rows))
      (is (= 20 (count rows)) "20-club league")
      (is (= 1 (:rank (first rows))))
      ;; points are non-increasing down the table
      (is (apply >= (map :points rows)))
      ;; 2019 champions were Flamengo
      (is (= "Flamengo" (:team (first rows)))))))

(deftest list-competitions-test
  (testing "Given the graph, When listing competitions, Then each has seasons and counts"
    (let [cs (q/list-competitions (h/graph))]
      (is (seq cs))
      (is (every? #(pos? (:matches %)) cs)))))

;; ---- 5. Statistical analysis ---------------------------------------------

(deftest league-stats-test
  (testing "Given Brasileirão matches, When aggregating, Then averages and rates are in plausible ranges"
    (let [s (q/league-stats (h/graph) {:competition "Brasileirão Série A"})]
      (is (pos? (:scored-matches s)))
      (is (< 1.5 (:avg-goals s) 4.0) "avg goals per match is plausible")
      (is (< 0.0 (:home-win-rate s) 1.0))
      (is (< 0.99 (+ (:home-win-rate s) (:away-win-rate s) (:draw-rate s)) 1.01)
          "rates sum to ~1"))))

(deftest biggest-wins-test
  (testing "Given match data, When finding biggest wins, Then they are sorted by margin descending"
    (let [ms (q/biggest-wins (h/graph) {:competition "Brasileirão Série A" :limit 5})]
      (is (= 5 (count ms)))
      (is (apply >= (map :margin ms)))
      (is (>= (:margin (first ms)) 4)))))
